internal/services/gateway/action/invitation.go

```diff
@@ -0,0 +1,245 @@
+	}
+
+	if req.UserRef == "" {
+		return nil, util.NewAPIError(util.ErrBadRequest, errors.Errorf("user id required"))
+	}
+	if req.OrganizationRef == "" {
+		return nil, util.NewAPIError(util.ErrBadRequest, errors.Errorf("organization id required"))
+	}
+	if req.Role == "" {
+		return nil, util.NewAPIError(util.ErrBadRequest, errors.Errorf("role is required"))
+	}
+
+	org, err := h.GetOrg(ctx, req.OrganizationRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get org %s:", req.OrganizationRef)
+	}
+
+	isOrgOwner, err := h.IsOrgOwner(ctx, org.ID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to determine ownership")
+	}
+	if !isOrgOwner {
+		return nil, util.NewAPIError(util.ErrForbidden, errors.Errorf("user not authorized"))
+	}
+
+	isOrgMember, err := h.IsOrgMember(ctx, req.UserRef, req.OrganizationRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to determine membership")
+	}
+	if isOrgMember {
+		return nil, util.NewAPIError(util.ErrForbidden, errors.Errorf("user is just an org member"))
+	}
+
+	creq := &csapitypes.CreateOrgInvitationRequest{
+		UserRef: req.UserRef,
+		Role:    req.Role,
+	}
+
+	h.log.Info().Msgf("creating org invitation")
+	orgInvitation, _, err := h.configstoreClient.CreateOrgInvitation(ctx, org.ID, creq)
+	if err != nil {
+		return nil, util.NewAPIError(util.KindFromRemoteError(err), errors.Wrapf(err, "failed to create org invitation"))
+	}
+	h.log.Info().Msgf("org invitation created, ID: %s", orgInvitation.ID)
+
+	return &OrgInvitationResponse{
+		OrgInvitation: orgInvitation,
+		Organization:  org,
+	}, nil
+}
+
+type OrgInvitationActionRequest struct {
+	OrgRef  string
+	UserRef string
+	Action  gwapitypes.OrgInvitationActionType `json:"action_type"`
+}
+
+func (h *ActionHandler) OrgInvitationAction(ctx context.Context, req *OrgInvitationActionRequest) error {
+	if !req.Action.IsValid() {
+		return errors.Errorf("action is not valid")
+	}
+
+	userID := common.CurrentUserID(ctx)
+	if userID == "" {
+		return util.NewAPIError(util.ErrBadRequest, errors.Errorf("user not authenticated"))
+	}
+
+	orgInvitation, _, err := h.configstoreClient.GetOrgInvitation(ctx, req.OrgRef, req.UserRef)
+	if err != nil {
+		return util.NewAPIError(util.KindFromRemoteError(err), errors.Wrapf(err, "failed to get org invitation"))
+	}
+	if orgInvitation == nil {
+		return util.NewAPIError(util.ErrBadRequest, errors.Errorf("invitation for org %s user %s not found", req.OrgRef, req.UserRef))
+	}
+
+	if userID != orgInvitation.UserID {
+		return errors.Errorf("user not authorized")
+	}
+
+	org, err := h.GetOrg(ctx, req.OrgRef)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get org %s:", req.OrgRef)
+	}
+	if org == nil {
+		return util.NewAPIError(util.ErrBadRequest, errors.Errorf("org %s not found", req.OrgRef))
+	}
+
+	switch req.Action {
+	case types.Accept:
+		_, _, err = h.configstoreClient.AddOrgMember(ctx, orgInvitation.OrganizationID, orgInvitation.UserID, orgInvitation.Role)
```

Comment:
You should not use configstore AddOrgMember since it'll add the user to regardless of invitation. You should add a configstore method to accept/deny invitation that will correctly enforce the invitation logic.